### PR TITLE
feat(auth): add data platform instance field resolver provider

### PR DESCRIPTION
**Pull request by amanda-her**
In this pull request the data platform instance field resolver has been added to return the platform instance of a resource. This is required for us to develop a custom authorizer using the plugin approach.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public FieldResolver getFieldResolver(ResourceSpec resourceSpec) {
+        return FieldResolver.getResolverFromFunction(resourceSpec, this::getDataPlatformInstance);
+    }
+
+    private FieldResolver.FieldValue getDataPlatformInstance(ResourceSpec resourceSpec) {
+        Urn entityUrn = UrnUtils.getUrn(resourceSpec.getResource());
+        // In the case that the entity is a platform instance, the associated platform instance entity is the instance itself
+        if (entityUrn.getEntityType().equals(DATA_PLATFORM_ENTITY_NAME)) {
```
Comment (sgomezvillamor) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java:44]: mabye?
```suggestion
        if (entityUrn.getEntityType().equals(DATA_PLATFORM_INSTANCE_ENTITY_NAME)) {
```

**Comment by sgomezvillamor** [comment]
Hi @jjoyce0510 ! At some moment we talked about the possibility of having a `DataPlatformInstanceFieldResolver` so we can easily fetch platform instance values in our custom authorizers (authorize plugins). Here it is the contrib!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Override
+    public FieldResolver getFieldResolver(ResourceSpec resourceSpec) {
+        return FieldResolver.getResolverFromFunction(resourceSpec, this::getDataPlatformInstance);
+    }
+
+    private FieldResolver.FieldValue getDataPlatformInstance(ResourceSpec resourceSpec) {
+        Urn entityUrn = UrnUtils.getUrn(resourceSpec.getResource());
+        // In the case that the entity is a platform instance, the associated platform instance entity is the instance itself
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java:43]: nice!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Objects;
+
+import static com.linkedin.metadata.Constants.*;
+
+/**
+ * Provides field resolver for domain given resourceSpec
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class DataPlatformInstanceFieldResolverProvider implements ResourceFieldResolverProvider {
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/DataPlatformInstanceFieldResolverProvider.java:26]: This looks great. If we can add a unit test for this provider we'll be good to merge!

**Review by jjoyce0510 — APPROVED** [review]
Overall LGTM. Let's just add one quick unit test then we are good to merge!

**Comment by amanda-her** [comment]
@jjoyce0510 Tests are already added!

**Pull request merged**